### PR TITLE
[experimental] An example usage of REST API

### PR DESCRIPTION
**Pull request by dmitri-perelman**
Retrieve the account sequence number via the REST API instead of going through the old json rpc client.
We should probably have a REST API client with a number of util functions, like retrieving the resource or event, sending a transaction, retrieving transaction result, etc.

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: diem-move/diem-framework/experimental/demo-cli/src/main.rs [hunk 1]
```diff
@@ -51,13 +53,10 @@ async fn main() -> Result<()> {
...
-    let client = BlockingClient::new(json_rpc_url);
+    let client = BlockingClient::new(json_rpc_url.clone());
 
-    let seq_num = client
-        .get_account(address)?
-        .into_inner()
-        .unwrap()
-        .sequence_number;
+    let seq_num =
+        request_utils::get_account_seq_num(json_rpc_url.as_str(), args.account_address).await?;
```
Comment (gregnazario) [inline diem-move/diem-framework/experimental/demo-cli/src/main.rs:59]: This feels really weird to create a client, and then get the sequence number via a util that bypasses the client.  I'd flip these.

I'd also rename request_utils to rest_request_utils

File: diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        get_account_resource_by_type(base_uri, account_address, DIEM_ACCOUNT_RESOURCE).await?;
+    for (k, v) in diem_account_resource
+        .as_object()
+        .ok_or_else(|| format_err!("Fail to parse resource {}", DIEM_ACCOUNT_RESOURCE))?
+        .iter()
+    {
+        if k == "sequence_number" {
+            return Ok(u64::from_str(v.as_str().unwrap()).unwrap());
+        }
+    }
```
Comment (gregnazario) [inline diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs:58]: I think what you're looking for is 

```
diem_account_resource
        .as_object()
        .ok_or_else(|| format_err!("Fail to parse resource {}", DIEM_ACCOUNT_RESOURCE))?
        .iter()
        .find(|(k,v)| k == "sequence_number")
        .ok_or("Resource sequence_number not found");
```

File: diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        .await?
+        .json::<Vec<AccountResource>>()
+        .await
+        .map_err(|e| format_err!("{}", e))?;
+    for resource_entry in acc_resources {
+        if resource_entry.type_ == resource_type {
+            return Ok(resource_entry.value);
+        }
+    }
+    bail!("Resource {} not found", resource_type);
```
Comment (gregnazario) [inline diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs:44]: Similar to the other one with the find on the iter.

**Review by gregnazario — APPROVED** [review]
Minor comments for the demo.  I would probably add some comment explaining that you're using REST other than just from the commit message

File: diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+use anyhow::{bail, format_err, Result};
+
+use serde::{Deserialize, Serialize};
+use std::str::FromStr;
+
+const DIEM_ACCOUNT_RESOURCE: &str = "0x1::DiemAccount::DiemAccount";
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct AccountResource {
```
Comment (xli) [inline diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs:12]: may use diem_api_types::MoveResource

File: diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    #[serde(rename(serialize = "type", deserialize = "type"))]
+    pub type_: String,
+    pub value: serde_json::Value,
```
Comment (xli) [inline diem-move/diem-framework/experimental/demo-cli/src/request_utils.rs:15]: Heads up, I'm landing the pr changes this field name to 'data', all others are same

**Comment by bmwill** [comment]
This PR https://github.com/diem/diem/pull/9823 introduces a new rest client which can handle some of the rest APIs. Its still WIP but should be basic enough for people to start using it for querying account resources and such.

**Pull request closed without merging**